File: Central/Assets/Scripts/Animations/AnimationSimple.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using System.Runtime.InteropServices;

namespace Animations
{
	/// <summary>
	/// Procedural on off animation
	/// </summary>
    [StructLayout(LayoutKind.Sequential, Pack = 1)]
    [System.Serializable]
	public class AnimationSimple
		: Animation
	{
		public AnimationType type { get; set; } = AnimationType.Simple;
		public byte padding_type { get; set; }
		public ushort duration { get; set; }
		public uint faceMask;
        public ushort colorIndex;
        public byte count;
        public byte fade;

        public AnimationInstance CreateInstance(DataSet.AnimationBits bits)
        {
            return new AnimationInstanceSimple(this, bits);
        }
	};

	/// <summary>
	/// Procedural on off animation instance data
	/// </summary>
	public class AnimationInstanceSimple
		: AnimationInstance
	{
        uint rgb = 0;

		public AnimationInstanceSimple(Animation animation, DataSet.AnimationBits bits)
            : base(animation, bits)
        {
        }

		public override void start(int _startTime, byte _remapFace, bool _loop)
        {
            base.start(_startTime, _remapFace, _loop);
            var preset = getPreset();
            rgb = animationBits.getColor32(preset.colorIndex);
        }

        public override int updateLEDs(int ms, int[] retIndices, uint[] retColors)
        {
            var preset = getPreset();

            // Compute color
            uint black = 0;
            uint color = 0;
            int period = preset.duration / preset.count;
            int fadeTime = period * preset.fade / (255 * 2);
            int onOffTime = (period - fadeTime * 2) / 2;
            int time = (ms - startTime) % period;

            if (time <= fadeTime) {
                // Ramp up
                color = ColorUtils.interpolateColors(black, 0, rgb, fadeTime, time);
            } else if (time <= fadeTime + onOffTime) {
                color = rgb;
            } else if (time <= fadeTime * 2 + onOffTime) {
                // Ramp down
                color = ColorUtils.interpolateColors(rgb, fadeTime + onOffTime, black, fadeTime * 2 + onOffTime, time);
            } else {
                color = black;
            }

            // Fill the indices and colors for the anim controller to know how to update leds
            int retCount = 0;
            for (int i = 0; i < 20; ++i) {
                if ((preset.faceMask & (1 << i)) != 0)
                {
                    retIndices[retCount] = i;
                    retColors[retCount] = color;
                    retCount++;
                }
            }
            return retCount;
        }

		public override int stop(int[] retIndices)
        {
            var preset = getPreset();
            int retCount = 0;
            for (int i = 0; i < 20; ++i) {
                if ((preset.faceMask & (1 << i)) != 0)
                {
                    retIndices[retCount] = i;
                    retCount++;
                }
            }
            return retCount;
        }

		public AnimationSimple getPreset()
        {
            return (AnimationSimple)animationPreset;
        }
	};
}

File: Firmware/src/animations/animation_simple.cpp
#include "animation_simple.h"
#include "utils/utils.h"
#include "config/board_config.h"

namespace Animations
{
	/// <summary>
	/// constructor for simple animations
	/// Needs to have an associated preset passed in
	/// </summary>
	AnimationInstanceSimple::AnimationInstanceSimple(const AnimationSimple* preset, const DataSet::AnimationBits* bits)
		: AnimationInstance(preset, bits) {
	}

	/// <summary>
	/// destructor
	/// </summary>
	AnimationInstanceSimple::~AnimationInstanceSimple() {
	}

	/// <summary>
	/// Small helper to return the expected size of the preset data
	/// </summary>
	int AnimationInstanceSimple::animationSize() const {
		return sizeof(AnimationSimple);
	}

	/// <summary>
	/// (re)Initializes the instance to animate leds. This can be called on a reused instance.
	/// </summary>
	void AnimationInstanceSimple::start(int _startTime, uint8_t _remapFace, bool _loop) {
		AnimationInstance::start(_startTime, _remapFace, _loop);
	}

	/// <summary>
	/// Computes the list of LEDs that need to be on, and what their intensities should be.
	/// </summary>
	/// <param name="ms">The animation time (in milliseconds)</param>
	/// <param name="retIndices">the return list of LED indices to fill, max size should be at least 21, the max number of leds</param>
	/// <param name="retColors">the return list of LED color to fill, max size should be at least 21, the max number of leds</param>
	/// <returns>The number of leds/intensities added to the return array</returns>
	int AnimationInstanceSimple::updateLEDs(int ms, int retIndices[], uint32_t retColors[]) {
        
        auto preset = getPreset();

        // Compute color
        uint32_t black = 0;
        uint32_t color = 0;
        int period = preset->duration / preset->count;
        int fadeTime = period * preset->fade / (255 * 2);
        int onOffTime = (period - fadeTime * 2) / 2;
        int time = (ms - startTime) % period;

        if (time <= fadeTime) {
            // Ramp up
            color = Utils::interpolateColors(black, 0, preset->color, fadeTime, time);
        } else if (time <= fadeTime + onOffTime) {
            color = preset->color;
        } else if (time <= fadeTime * 2 + onOffTime) {
            // Ramp down
            color = Utils::interpolateColors(preset->color, fadeTime + onOffTime, black, fadeTime * 2 + onOffTime, time);
        } else {
            color = black;
        }

        // Fill the indices and colors for the anim controller to know how to update leds
        int retCount = 0;
        for (int i = 0; i < 20; ++i) {
            if ((preset->faceMask & (1 << i)) != 0)
            {
                retIndices[retCount] = i;
                retColors[retCount] = color;
                retCount++;
            }
        }
        return retCount;
	}

	/// <summary>
	/// Clear all LEDs controlled by this animation, for instance when the anim gets interrupted.
	/// </summary>
	int AnimationInstanceSimple::stop(int retIndices[]) {
        auto preset = getPreset();
        int retCount = 0;
        for (int i = 0; i < 20; ++i) {
            if ((preset->faceMask & (1 << i)) != 0)
            {
                retIndices[retCount] = i;
                retCount++;
            }
        }
        return retCount;
	}

	const AnimationSimple* AnimationInstanceSimple::getPreset() const {
        return static_cast<const AnimationSimple*>(animationPreset);
    }

}
